src/Orleans.Core/Diagnostics/ActivityPropagationGrainCallFilter.cs

```diff
@@ -0,0 +1,108 @@
+            if (activity is not null)
+            {
+                RequestContext.Set(TraceParentHeaderName, activity.Id);
+            }
+
+            try
+            {
+                await context.Invoke();
+                if (activity is not null && activity.IsAllDataRequested)
+                {
+                    activity.SetTag("status", "Ok");
+                }
+            }
+            catch (Exception e)
+            {
+                if (activity is not null && activity.IsAllDataRequested)
+                {
+                    // exception attributes from https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/exceptions.md
+                    activity.SetTag("exception.type", e.GetType().FullName);
+                    activity.SetTag("exception.message", e.Message);
+                    activity.SetTag("exception.stacktrace", e.StackTrace);
+                    activity.SetTag("exception.escaped", true);
+                    activity.SetTag("status", "Error");
+                }
+                throw;
+            }
+        }
+    }
+
+    internal class ActivityPropagationOutgoingGrainCallFilter : ActivityPropagationGrainCallFilter, IOutgoingGrainCallFilter
+    {
+        public Task Invoke(IOutgoingGrainCallContext context)
+        {
+            if (Activity.Current != null)
+            {
+                return ProcessCurrentActivity(context); // Copy existing activity to RequestContext
+            }
+            return ProcessNewActivity(context, ActivityNameOut, ActivityKind.Client, new ActivityContext()); // Create new activity
+        }
+
+        private static Task ProcessCurrentActivity(IOutgoingGrainCallContext context)
+        {
+            var currentActivity = Activity.Current;
+
+            if (currentActivity is not null &&
+                currentActivity.IdFormat == ActivityIdFormat.W3C)
+            {
+                RequestContext.Set(TraceParentHeaderName, currentActivity.Id);
+                if (currentActivity.TraceStateString is not null)
+                    RequestContext.Set(TraceStateHeaderName, currentActivity.TraceStateString);
+            }
+
+            return context.Invoke();
+        }
+    }
+
+    internal class ActivityPropagationIncomingGrainCallFilter : ActivityPropagationGrainCallFilter, IIncomingGrainCallFilter
+    {
+        public Task Invoke(IIncomingGrainCallContext context)
+        {
+            // Create activity from context directly
+            var traceParent = RequestContext.Get(TraceParentHeaderName) as string;
+            var traceState = RequestContext.Get(TraceStateHeaderName) as string;
+            var parentContext = new ActivityContext();
+
+            if (traceParent is not null)
+            {
+                parentContext = ActivityContext.Parse(traceParent, traceState);
+            }
```

Comment:
I would replace this logic and rely on the configured `DistributedContextPropagator` instead. I'm not familiar with Orleans, but in ASP.NET Core we use the propagator registered in DI and if not present, we fall back to the globally configured `DistributedContextPropagator.Current`

Reply:
Also, was it intentional not to propagate Baggage? Or was it an omission?

Reply:
DistributedContextPropagator is NET 6.0 only. I'm not sure Orleans 4.0 should be constrained to that, unless there's plans to backport it to 5.0

Reply:
You may try to reference the latest Diagnostic Source package https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/6.0.0-preview.7.21377.19 which should be supporting the propagators and should work in .NET 5.0 too.

Reply:
@tarekgh we have the same discussion about v5 [packages](https://github.com/dotnet/orleans/pull/6853#issuecomment-742698992). I can't just upgrade packages especially to preview versions. I think we should refactor this code after orleans upgrades all nuget references to 6.0.

Reply:
@tarekgh Oh, that would be good if it works :)